### DISCUSSION TITLE
Makerfabs ESP32-S3 with 480x320 parallel 3.5 example

### DISCUSSION THREAD

**Discussion by ammaree** [sukesh-ak/ESP32-TUX#5] [Q&A] [answered]
As per Issue #385 at lovyan03/LovyanGFX I am looking for some help. Background is as follows:
Makerfabs ESP32S3 3.5" 480x320 16bit parallel board

Need to integrate into an existing (large) "C" codebase using ESP IDF v5.x
Prefer to keep things simple, use only LGFX without LVGL complications
Initially only need to display about 2400 (60x40) text characters, updated every 100-250mS
Need touch just to turn on/off the display activity.
At later stage will actually convert into a proper GUI

Looking for an example, simplest possible, to
a) initialise the ILI9488 and FT6236
b) clear the screen
c) display a text buffer
d) switch on/off (set a flag) on touch

Any suggestions

**Comment by sukesh-ak** [comment]
LovyanGFX configuration required for Makerfabs ESP32S3 3.5" 480x320 16bit parallel board is here for both Display & Touch.
Its tested since I just got the board from Makerfabs.
https://github.com/sukesh-ak/ESP32-TUX/blob/master/main/devices/conf_Makerfabs_S3_PTFT.h

You cannot use LovyanGFX (C++) with C code. The only thing you need to do is to use extern "C" for your app_main and rename your `main.c` to `main.cpp`. Everything else can be `.c` and the compiler will take care of it.

```c++
extern "C" void app_main(void)
{

}

```

**Comment by sukesh-ak** [comment]
The above config works for IDF 4.x
For making it work for IDF 5.x you can use LovyanGFX `develop branch` (till its released) which has the IDF 5.x fixes.

**Comment by ammaree** [comment]
Everything you have mentioned I have done. The compile goes through with 3 deprecation warnings for driver/definition issuesBUT it then fails in the linker stage.

I am using a 100% copy of the simplest test example from the library, with setup() and loop() code called, at the appropriate stages, from my task.

Linker errors are as follows:
/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.literal.spi_master_init_driver+0x40): undefined reference to `spi_hal_init'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.literal.spi_master_init_driver+0x4c): undefined reference to `spi_hal_deinit'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.6.literal+0x0): undefined reference to `spi_hal_fetch_result'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.0.literal+0x0): undefined reference to `spi_hal_setup_device'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.5.literal+0x0): undefined reference to `spi_hal_setup_trans'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.5.literal+0x4): undefined reference to `spi_hal_prepare_data'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.5.literal+0x8): undefined reference to `spi_hal_user_start'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.iram1.7.literal+0x18): undefined reference to `spi_hal_usr_is_done'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj):(.literal.spi_bus_add_device+0x48): undefined reference to `spi_hal_cal_clock_conf'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_master_init_driver':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:241: undefined reference to `spi_hal_init'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:255: undefined reference to `spi_hal_deinit'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_master_deinit_driver':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:277: undefined reference to `spi_hal_deinit'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_post_trans':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:604: undefined reference to `spi_hal_fetch_result'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_setup_device':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:502: undefined reference to `spi_hal_setup_device'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_new_trans':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:589: undefined reference to `spi_hal_setup_trans'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:590: undefined reference to `spi_hal_prepare_data'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:595: undefined reference to `spi_hal_user_start'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_intr':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:619: undefined reference to `spi_hal_usr_is_done'

/Users/andremaree/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: esp-idf/driver/libdriver.a(spi_master.c.obj): in function `spi_bus_add_device':

/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/spi_master.c:371: undefined reference to `spi_hal_cal_clock_conf`

**Comment by sukesh-ak** [comment]
Looks like these are related to IDF 5.x breaking changes.

Which version are you using? IDF 5.0 or IDF 5.1? Both are unreleased.

You can check this for SPI changes in IDF 5.x as a start since the file name has changed.
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/migration-guides/release-5.x/5.0/peripherals.html#spi

_This is the reason I didn't jump on IDF 5.x and waiting till its released. Unreleased versions are always a moving target._

**Comment by sukesh-ak** [comment]
@ammaree Also did you switch to `develop` branch in LovyanGFX ?

**Comment by ammaree** [comment]
Using ESP IDF Master, so effectively RC5.0
Yes, switched to Lovyan  develop branch

On 2022/11/30, 17:38, "Sukesh Ashok Kumar" ***@***.******@***.***>> wrote:


@ammaree<https://github.com/ammaree> Also did you switch to develop branch in LovyanGFX ?

**Comment by sukesh-ak** [comment]
Btw, you said you have Parallel version of the display. The error you mentioned says SPI, so something is not right.

Master is 5.1 I believe. 
5.0 is on the Release branch [here](https://github.com/espressif/esp-idf/tree/release/v5.0)

**Reply by ammaree** [reply]
I have a number of different modules so with others without display support.
We are developing support for a range of hardware. 
To simplify the setup I reverted back to an Espressif WROVER DevKit v4.1 that is directly supported in Lovyan.

`/*
 * gui_lovyan_demo.cpp
 * Copyright (c) 2022 Andre M. Maree / KSS Technologies (Pty) Ltd.
 */

#include "main.h"
#if	(cmakeGUI == 3)

#define LGFX_AUTODETECT
#define LGFX_USE_V1
#include <LovyanGFX.hpp>

#include "hal_variables.h"
#include "systiming.h"

// #################################### Build macros ###############################################

#define	debugFLAG					0xF000

#define	debugTIMING					(debugFLAG_GLOBAL & debugFLAG & 0x1000)
#define	debugTRACK					(debugFLAG_GLOBAL & debugFLAG & 0x2000)
#define	debugPARAM					(debugFLAG_GLOBAL & debugFLAG & 0x4000)
#define	debugRESULT					(debugFLAG_GLOBAL & debugFLAG & 0x8000)

// ################################# Device definitions ############################################

#define	guiPRIORITY					2
#define	guiSTACK_SIZE				(configMINIMAL_STACK_SIZE + 2048 + (flagSTACK * 512))
#define guiINTERVAL_MS				100

// ####################################### Private variables #######################################

StaticTask_t ttsGUI;
StackType_t tsbGUI[guiSTACK_SIZE] = { 0 };

LGFX lcd;
LGFX_Sprite canvas;

static constexpr char text[] = "Hello world ! こんにちは世界！ this is long long string sample. 寿限無、寿限無、五劫の擦り切れ、海砂利水魚の、水行末・雲来末・風来末、喰う寝る処に住む処、藪ら柑子の藪柑子、パイポ・パイポ・パイポのシューリンガン、シューリンガンのグーリンダイ、グーリンダイのポンポコピーのポンポコナの、長久命の長助";
static constexpr size_t textlen = sizeof(text) / sizeof(text[0]);
size_t textpos = 0;

void vGuiInit(void) {
  lcd.init();
  if (lcd.width() < lcd.height())
	  lcd.setRotation(lcd.getRotation() ^ 1);
  canvas.setColorDepth(8);
  canvas.setFont(&fonts::lgfxJapanMinchoP_32);
  canvas.setTextWrap(false);
  canvas.createSprite(lcd.width() + 36, 36);
}

void vGuiUpdate(void) {
  int32_t cursor_x = canvas.getCursorX() - 1;
  if (cursor_x <= 0) {
    textpos = 0;
    cursor_x = lcd.width();
  }

  canvas.setCursor(cursor_x, 0);
  canvas.scroll(-1, 0);
  while (textpos < textlen && cursor_x <= lcd.width()) {
    canvas.print(text[textpos++]);
    cursor_x = canvas.getCursorX();
  }
  canvas.pushSprite(&lcd, 0, 0);
}

void vGuiRefresh(void) {}

void vTaskGUI(void * pVoid) {
	IF_SYSTIMER_INIT(debugTIMING, stGUI0, stMILLIS, "GUI0", 3, 30);
	IF_SYSTIMER_INIT(debugTIMING, stGUI1, stMILLIS, "GUI1", 20, 100);

	vGuiInit();
	vTaskSetThreadLocalStoragePointer(NULL, 1, (void *)taskGUI_MASK);
	TickType_t GUI_LWtime = xTaskGetTickCount();
	xRtosSetStateRUN(taskGUI_MASK);

    while(bRtosVerifyState(taskGUI_MASK)) {
    	IF_SYSTIMER_START(debugTIMING, stGUI0);
    	vGuiUpdate();
    	IF_SYSTIMER_STOP(debugTIMING, stGUI0);

    	IF_SYSTIMER_START(debugTIMING, stGUI1);
    	vGuiRefresh();
    	IF_SYSTIMER_STOP(debugTIMING, stGUI1);

		vTaskDelayUntil(&GUI_LWtime, pdMS_TO_TICKS(guiINTERVAL_MS));
	}
	vRtosTaskDelete(NULL);
}

void vTaskGUI_Init(void * pvPara) {
	xRtosTaskCreateStatic(vTaskGUI, "lovyan", guiSTACK_SIZE, pvPara, guiPRIORITY, tsbGUI, &ttsGUI, tskNO_AFFINITY);
}
#endif
`

**Reply by sukesh-ak** [reply]
I will get you a simple sample with LovyanGFX working for that board.
Then you can go from there.

**Comment by sukesh-ak** [comment, marked as answer]
@ammaree 
Here is a sample you can copy past in Arduino or use in IDF. 
Draw using your finger on the screen.

LovyanGFX only (no LVGL)
https://gist.github.com/sukesh-ak/7bcafd2227a2c2fa727208f4ffa47308

**Comment by ammaree** [comment]
Thanks, need to swing build environment back to Makerfabs config to build this, will test bit later tonight

**Reply by sukesh-ak** [reply]
You can also open a new window in Arduino and copy/paste and flash.

**Reply by ammaree** [reply]
We do not use Arduino at all, not even installed.

**Reply by sukesh-ak** [reply]
Cool. I have it installed and use only for quick testing.

**Comment by ammaree** [comment]
My link problem, using your email code, remains the same when tested in the IDF v5 environment. Testing in Arduino might be different, not sure if Arduino is using latest from IDF.
 
ESP IDF v5.0 has just been officially released as a stable version, maybe somebody can convince LGFX maintainers to complete their migration.

Possibly get it ready and integrated as an official component at https://components.espressif.com

I can only hope....

**Reply by sukesh-ak** [reply]
> My link problem, using your email code, remains the same when tested in the IDF v5 environment.

Not sure what you mean. IDF 5.0 related issues need to be fixed by you. It works fine with IDF 4.x already. Autodetect might not be working but the code I shared in the `gist` should work with IDF 5.x too

> ESP IDF v5.0 has just been officially released as a stable version, maybe somebody can convince LGFX maintainers to complete their migration.

Thanks for the update. I will upgrade my work now to IDF 5.0

**Comment by ammaree** [comment]
What I mean is that there are compile error related to LGFX includes and symbols that need to be changed/fixed for IDF v5.0.
ALL my code has been migrated to v5.0, only problem outstanding is the GUI using LGFX

Example:
`[787/945] Building CXX object esp-idf/lovyan/CMakeFiles/__idf_lovyan.dir/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp.objIn file included from /Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp:31:
/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/deprecated/driver/dac.h:16:2: warning: #warning "The legacy DAC driver is deprecated, please use `driver/dac_oneshot.h`, `driver/dac_cosine.h` or `driver/dac_continuous.h` instead" [-Wcpp]
   16 | #warning "The legacy DAC driver is deprecated, please use `driver/dac_oneshot.h`, `driver/dac_cosine.h` or `driver/dac_continuous.h` instead"
      |  ^~~~~~~
/Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp: In member function 'virtual bool lgfx::v1::Panel_CVBS::init(bool)':
/Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp:930:25: warning: 'DAC_CHANNEL_1' is deprecated: please use 'DAC_CHAN_0' instead [-Wdeprecated-declarations]
  930 |       dac_output_enable(DAC_CHANNEL_1); // for GPIO 25
      |                         ^~~~~~~~~~~~~
In file included from /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/deprecated/driver/dac_types_legacy.h:14,
                 from /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/deprecated/driver/dac.h:13,
                 from /Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp:31:
/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/hal/include/hal/dac_types.h:10:5: note: declared here
   10 |     DAC_CHANNEL_1 __attribute__((deprecated("please use 'DAC_CHAN_0' instead"))) = 0,    /*!< Alias of 'DAC_CHAN_0', now the channel index start from '0' */
      |     ^~~~~~~~~~~~~
/Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp:933:25: warning: 'DAC_CHANNEL_2' is deprecated: please use 'DAC_CHAN_1' instead [-Wdeprecated-declarations]
  933 |       dac_output_enable(DAC_CHANNEL_2); // for GPIO 26
      |                         ^~~~~~~~~~~~~
In file included from /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/deprecated/driver/dac_types_legacy.h:14,
                 from /Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/driver/deprecated/driver/dac.h:13,
                 from /Users/andremaree/Dropbox/devs/ws/z-repo/lovyan/src/lgfx/v1/platforms/esp32/Panel_CVBS.cpp:31:
/Users/andremaree/Dropbox/devs/ws/z-sdk/esp-idf/components/hal/include/hal/dac_types.h:11:5: note: declared here
   11 |     DAC_CHANNEL_2 __attribute__((deprecated("please use 'DAC_CHAN_1' instead"))) = 1,    /*!< Alias of 'DAC_CHAN_1', now the channel index start from '0' */
      |     ^~~~~~~~~~~~~
`

**Reply by sukesh-ak** [reply]
Did you to switch to `develop` branch of LovyanGFX?

**Reply by sukesh-ak** [reply]
Its discussed here https://github.com/lovyan03/LovyanGFX/issues/282

**Reply by ammaree** [reply]
As per my message 2 days ago, I switched to develop branch at the start...

**Reply by sukesh-ak** [reply]
I just upgraded my project to IDF 5.0

LovyanGFX with IDF 5.0 & Parallel drivers have a bug. SPI works fine though.
You can follow it [here](https://github.com/lovyan03/LovyanGFX/issues/322)

**Reply by sukesh-ak** [reply]
If you want to work around the bug you can follow the instructions [here](https://github.com/sukesh-ak/ESP32-TUX/issues/6)

SPI works fine. With the above work around, Parallel 8/16bit works as well.

**Reply by sukesh-ak** [reply]
@ammaree 
Project is updated with LovyanGFX Parallel devices related fix.
Also check the readme for updated instructions including the link for Wi-Fi provisioning app.
